packages/rich-text/src/component/use-input-and-selection.js

```diff
@@ -0,0 +1,368 @@
+				return;
+			}
+
+			if ( disabled ) {
+				return;
+			}
+
+			// In case of a keyboard event, ignore selection changes during
+			// composition.
+			if ( isComposing ) {
+				return;
+			}
+
+			const { start, end, text } = createRecord();
+			const oldRecord = record.current;
+
+			// Fallback mechanism for IE11, which doesn't support the input event.
+			// Any input results in a selection change.
+			if ( text !== oldRecord.text ) {
+				onInput();
+				return;
+			}
+
+			if ( start === oldRecord.start && end === oldRecord.end ) {
+				// Sometimes the browser may set the selection on the placeholder
+				// element, in which case the caret is not visible. We need to set
+				// the caret before the placeholder if that's the case.
+				if ( oldRecord.text.length === 0 && start === 0 ) {
+					fixPlaceholderSelection( defaultView );
+				}
+
+				return;
+			}
+
+			const newValue = {
+				...oldRecord,
+				start,
+				end,
+				// _newActiveFormats may be set on arrow key navigation to control
+				// the right boundary position. If undefined, getActiveFormats will
+				// give the active formats according to the browser.
+				activeFormats: oldRecord._newActiveFormats,
+				_newActiveFormats: undefined,
+			};
+
+			const newActiveFormats = getActiveFormats(
+				newValue,
+				EMPTY_ACTIVE_FORMATS
+			);
+
+			// Update the value with the new active formats.
+			newValue.activeFormats = newActiveFormats;
+
+			if ( ! isCaretWithinFormattedText && newActiveFormats.length ) {
+				onEnterFormattedText();
+			} else if (
+				isCaretWithinFormattedText &&
+				! newActiveFormats.length
+			) {
+				onExitFormattedText();
+			}
+
+			// It is important that the internal value is updated first,
+			// otherwise the value will be wrong on render!
+			record.current = newValue;
+			applyRecord( newValue, { domOnly: true } );
+			onSelectionChange( start, end );
+			setActiveFormats( newActiveFormats );
+		}
+
+		function onCompositionStart() {
+			isComposing = true;
+			// Do not update the selection when characters are being composed as
+			// this rerenders the component and might distroy internal browser
+			// editing state.
+			ownerDocument.removeEventListener(
+				'selectionchange',
+				handleSelectionChange
+			);
+		}
+
+		function onCompositionEnd() {
+			isComposing = false;
+			// Ensure the value is up-to-date for browsers that don't emit a final
+			// input event after composition.
+			onInput( { inputType: 'insertText' } );
+			// Tracking selection changes can be resumed.
+			ownerDocument.addEventListener(
+				'selectionchange',
+				handleSelectionChange
+			);
+		}
+
+		function onFocus() {
+			const {
+				record,
+				isSelected,
+				onSelectionChange,
+				setActiveFormats,
+			} = propsRef.current;
+
+			if ( ! isSelected ) {
+				// We know for certain that on focus, the old selection is invalid.
+				// It will be recalculated on the next mouseup, keyup, or touchend
+				// event.
```

Comment:
Do you remember why we have this comment @ellatrix ?

Is there a way to get the correct "start, end" without waiting for the next event (for performance reasons).
Also, would `0` be a better default than `undefined` if not?